src/ppx/metrics.cpp

```diff
@@ -0,0 +1,219 @@
+    if (entriesCount == 0) {
+        return statistics;
+    }
+
+    std::vector<TimeSeriesEntry> sorted = mTimeSeries;
+    std::sort(
+        sorted.begin(), sorted.end(), [](const TimeSeriesEntry& lhs, const TimeSeriesEntry& rhs) {
+            return lhs.value < rhs.value;
+        });
+
+    if (entriesCount % 2 == 0) {
+        const size_t medianIndex = entriesCount / 2;
+        PPX_ASSERT_MSG(medianIndex > 0, "Unexpected median index");
+        statistics.median = (sorted[medianIndex - 1].value + sorted[medianIndex].value) * 0.5;
+    }
+    else {
+        const size_t medianIndex = entriesCount / 2;
+        statistics.median        = sorted[medianIndex].value;
+    }
+
+    {
+        double squareDiffSum = 0.0;
+        for (auto entry : mTimeSeries) {
+            const double diff = entry.value - mBasicStatistics.average;
+            squareDiffSum += (diff * diff);
+        }
+        const double variance        = squareDiffSum / entriesCount;
+        statistics.standardDeviation = sqrt(variance);
+    }
+
+    const size_t percentileIndex90 = entriesCount * 90 / 100;
+    statistics.percentile90        = sorted[percentileIndex90].value;
+    const size_t percentileIndex95 = entriesCount * 95 / 100;
+    statistics.percentile95        = sorted[percentileIndex95].value;
+    const size_t percentileIndex99 = entriesCount * 99 / 100;
+    statistics.percentile99        = sorted[percentileIndex99].value;
+
+    return statistics;
+}
+
+void MetricGauge::UpdateBasicStatistics(double seconds, double value)
+{
+    mAccumulatedValue += value;
+
+    mBasicStatistics.min = std::min(mBasicStatistics.min, value);
+    mBasicStatistics.max = std::max(mBasicStatistics.max, value);
+
+    const size_t entriesCount = mTimeSeries.size();
+    mBasicStatistics.average  = mAccumulatedValue / entriesCount;
+    if (entriesCount > 1) {
+        const double timeSpan      = mTimeSeries.back().seconds - mTimeSeries.front().seconds;
+        mBasicStatistics.timeRatio = mAccumulatedValue / timeSpan;
+    }
+    else {
+        mBasicStatistics.timeRatio = mTimeSeries.front().value;
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+MetricCounter::MetricCounter(const MetricMetadata& metadata)
+    : mMetadata(metadata), mCounter(0U)
+{
+}
+
+MetricCounter::~MetricCounter()
+{
+}
+
+uint64_t MetricCounter::Increment(uint64_t add)
+{
+    mCounter += add;
+    return mCounter;
+}
+
+uint64_t MetricCounter::Get() const
+{
+    return mCounter;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+Run::Run(const char* name)
+    : mName(name)
+{
+    PPX_ASSERT_MSG(name != nullptr, "A run name cannot be null");
+}
+
+Run::~Run()
+{
+    for (auto [name, metric] : mGauges) {
+        delete metric;
+    }
+    mGauges.clear();
+    for (auto [name, metric] : mCounters) {
+        delete metric;
+    }
+    mCounters.clear();
+}
+
+void Run::AddMetric(MetricGauge* metric)
```

Comment:
I prefer the non-templated funtions. If we ever add another type, it will break at compile time 100% of time. It is straightforward, so I like it better. :)

Reply:
Fair